spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SerializableTableWithSize.java

```diff
@@ -33,8 +33,9 @@
  *
  * <p>This class also implements AutoCloseable to avoid leaking resources upon broadcasting.
  * Broadcast variables are destroyed and cleaned up on the driver and executors once they are
- * garbage collected on the driver. The implementation ensures only resources used by copies of the
- * main table are released.
+ * garbage collected on the driver. The implementation should avoid closing deserialized copies of
+ * shared resources like FileIO, as they may use a shared connection pool. Shutting down the pool
```

Comment:
which FileIO implementation shares the same connection pool across instances?

Reply:
SerializedTable (which is broadcast to the different instances) can get cleaned up/persisted to disk and this leads to the connection pool getting closed, which might still be in use. See also:

- https://github.com/apache/iceberg/issues/12046#issuecomment-2621663034
- https://github.com/apache/iceberg/pull/12129#issue-2818297163

Reply:
I believe the issue is specific to `S3FileIO` and I thought that each `S3FileIO` instance uses its own and dedicated `ApacheHttpClient` (because we set it up as described in https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/http-configuration-apache.html), which would create a new `S3Client` instance and therefore a new `ApacheHttpClient` after a Table got broadcasted or is that not the case?

Reply:
I think that's generally fine (each instance uses it's own dedicated client), but I rather think the issue is something like the following (sorry, it's been a while since I looked into this):

- SerializableTable is broadcast to the executor
- The IO object is used from the serializable table.
- At some point, Spark requests the object to be persisted to disk/block released. During this process the IO is closed.
- The IO object is still being used elsewhere in the executor and, when it is used *again*, this results in the failure.

Reply:
Hi @nastra, thanks for the review! You are right. The issue happen when using `ApacheHttpClient`, the issue occurs when using `ApacheHttpClient` because all `ApacheHttpClient` share the same connection pool instance (as seen in the [referenced code](https://github.com/apache/httpcomponents-client/blob/master/httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManager.java#L262)). That’s why it shouldn’t be closed, even during GC at least when `ApacheHttpClient` been used.